Quaternion Algebra Fractional Ideal improvements - equivalence and reduced bases

jtcc2 opened the pull request:
Includes the following changes
1. Added `.is_definite()` to quaternion algebras, which performs a check on the invariants.
2. Added `.reduced_basis()` to quaternion fractional ideals which uses LLL to find the basis of shortest elements
3. Modified quaternion fractional ideal `.is_equivalent` to include a `side` parameter which can be set to `'left'` or `'right'` to  do left-ideal or right-ideal equivalence, and a `certificate` parameter which if true, returns the element a such that I = J*a. Throws a `NotImplementedError` for indefinite quaternion algebras which fixes #37080 . Backwards compatible.
4. Also added `.is_left_equivalent` and `.is_right_equivalent`.
5. Added `.is_principal` to quaternion fractional ideals of rank 4 in definite quaternion algebras. Includes `certificate` parameter for finding generator. Finds the shortest norm element in the lattice and checks its norm.
6. Fixes #37337, bug in quaternion order `.isomorphism_to` where it cannot find the isomorphism between $\mathcal{O}$ and $\alpha^{-1} \mathcal{O} \alpha$ if $nrd(\alpha)$ is a ramified prime times a square. Fixed this by repeating the previous `isomorphism_to` method multiple times on inputs $\mathcal{O}$ and $\alpha^{-1} \mathcal{O} \alpha$ with $\alpha = 1, i, j, k$. Throws a `NotImplementedError` in cases not covered.

@S17A05 @Jonathke

#sd123

jtcc2 committed:
Added QuaternionAlgebra.is_definite(), Quaternion Ideal reduced_basis(), and is_equivalent certificate

jtcc2 committed:
Improved definiteness check for reduced_basis

jtcc2 committed:
reduced_basis must have full rank ideal

jtcc2 committed:
Removed whitespace

jtcc2 committed:
Added .is_principal

jtcc2 committed:
Fixed codestyle

jtcc2 committed:
Fixed codestyle

jtcc2 committed:
fix

jtcc2 committed:
fix

jtcc2 committed:
Fixed doctest

S17A05 committed:
Small modifications

- Removed greek letter alpha in docstrings in hopes of this fixing infinite build loops
- Updated `.is_definite()` to fit PR #37173 (up to the reference to Voight's book)
- Other small modifications of docstrings, comments and error warnings
- Slightly cleaned up code with respect to intermediately defined variables

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2477,6 +2500,43 @@ def basis_matrix(self):
+
+        if not self.quaternion_algebra().is_definite():
+            if not self.quadratic_form().is_positive_definite():
+                raise TypeError("The quaternion algebra must be definite")
```
Aren't these two tests equivalent? Were they duplicated by accident?

S17A05 replied:
Yes, these two should be equivalent - I believe the latter test was originally put in place for the method to also work for "lattices" of non-maximal rank, until it came up that `.gram_matrix()` needs a full rank lattice anyway. I'll remove the second test.

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
         """
-        # shorthand: let I be self
-        if not isinstance(self, QuaternionFractionalIdeal_rational):
+        if not isinstance(J, QuaternionFractionalIdeal_rational):
```
I'd say in this case we should probably throw a `TypeError`. Returning `False` is arguably correct, but I can't see any situation in which users will rely on this behaviour wilfully, hence failing visibly is better.

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2845,9 +2905,11 @@ def multiply_by_conjugate(self, J):
-    def is_equivalent(self, J, B=10) -> bool:
-        """
-        Return ``True`` if ``self`` and ``J`` are equivalent as right ideals.
+    def is_equivalent(self, J, B=10, certificate=False, side=None):
```
Personally I would simply deprecate this method in favor of the two new explicitly named one-sided methods: Having it default to `.is_right_equivalent()` is arbitrary and ambiguous.

S17A05 replied:
So should I add a deprecation warning for now to set it up for future removal, or should I remove it immediately?

yyyyx4 replied:
According to The Rules, we're only allowed to remove things once they've been deprecated for one year, so it has to be the former: Add a deprecation warning. (If you agree, that is: I don't have a strong opinion about this either way.)

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+
+        if not self.quaternion_algebra().is_definite():
+            raise NotImplementedError('equivalence test of ideals not implemented'
+                                    'for indefinite quaternion algebras')
```
```suggestion
            raise NotImplementedError('equivalence test of ideals not implemented'
                                    ' for indefinite quaternion algebras')
```

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+
+        # Use Pari's qfminim with flag = 1 to find alpha with N(alpha) = N(I)
+        _,v = self.quadratic_form().__pari__().qfminim(None, None, 1)
+        return True, sum(ZZ(c)*g for c,g in zip(v, self.basis()))
```
Same here re: the `.minimal_element()` method from #34976.

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+                                    'implemented in indefinite'
+                                    'quaternion algebras')
+
+        if len(self.basis()) < 4:
```
Can this fail?

S17A05 replied:
On faulty input, I believe, since calling `B.ideal([i,j])` for a quaternion algebra `B` gives a "fractional ideal", but it really is just a two-dimensional sublattice with basis `(i,j)`.

yyyyx4 replied:
Interesting, I was expecting the constructor for fractional ideals to check the rank. That should probably be fixed eventually, but it doesn't necessarily have to be done here and now.

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+        if not self.quaternion_algebra().is_definite():
+            raise NotImplementedError('principality test not'
+                                    'implemented in indefinite'
+                                    'quaternion algebras')
```
```suggestion
            raise NotImplementedError('principality test not'
                                    ' implemented in indefinite'
                                    ' quaternion algebras')
```

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+
+        # Use Pari's qfminim with flag = 1 to find alpha as above
+        _,v = IJbar.quadratic_form().__pari__().qfminim(None, None, 1)
+        alpha = sum(ZZ(c)*g for c,g in zip(v, IJbar.basis()))
```
This one could be `alpha = IJbar.minimal_element()` once #34976 is merged, but pulling that in as a dependency seems a bit overkill just to deduplicate these two lines. Maybe add a TODO so that eventually someone can replace it later?

S17A05 replied:
Since it might be a good idea to test definiteness with this PR's `.is_definite()`-method as well, it might actually be a good idea to put a dependency on this PR to yours (see also [this review comment](https://github.com/sagemath/sage/pull/34976#discussion_r1476246835)).

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+        # As explained by Pizer, we now only need to rescale alpha by 1/N(J)
+        return True, alpha * (1/J.norm())
+
+    def is_principal(self, certificate=False):
```
This method duplicates quite some code from the equivalence tests. Perhaps the equivalence tests could/should just end up calling `.is_principal()` on `IJbar` instead of repeating all the steps on that special case?

S17A05 committed:
Implemented first batch of reviewer feedback

- Added deprecation warning for `.is_equivalent`
- Removed unnecessary call of `.is_positive_definite()` in `.reduced_basis()`
- Reduced `.is_right_equivalent` to `.is_principal` with suitable rescaling
- Fixed some error warnings
- Modified docstrings, moved examples to non-deprecated methods and randomized test of `.is_principal`

S17A05 committed:
Fix lint issues

S17A05 committed:
(Attempt to) Fix deprecation failure

S17A05 committed:
Replace is_equivalent in BrandtModule

S17A05 committed:
Fully integrated #34976

- Modified definiteness check in `.minimal_element()` and `.isomorphism_to`; also added check of the base field in the latter method
- Reduced search for a generator in `.is_principal()` to a call to `.minimal_element()`

S17A05 committed:
Added check for fractional ideal basis length, removed "local" checks

yyyyx4 committed:
Merge branch 'public/isomorphisms_between_quaternion_orders'

S17A05 committed:
Merge branch 'develop' into develop

S17A05 committed:
Address issue #37337

This is not a fix for the issue - it just adapts the error warning to be more ambiguous, as suggested in https://github.com/sagemath/sage/pull/34976#issuecomment-1937007360

jtcc2 committed:
Fixed isomorphism_to bug

jtcc2 committed:
Fixed lint

jtcc2 committed:
Added tests

jtcc2 committed:
Added tests

jtcc2 committed:
Small correction

S17A05 committed:
Merge branch 'develop' into develop

S17A05 committed:
Merge branch 'develop' into develop

S17A05 committed:
Merge branch 'develop' into develop

yyyyx4 reviewed (approved):
Looks good to me now, thanks!
